subvolumegroup: add new crd (#9384)
rook/rook: Documentation/ceph-fs-subvolumegroup.md

---

> title: CephFilesystemSubVolumeGroup CRD

Reviewer: Might need to truncate the name in the TOC like we did for other CRD names.
```suggestion
title: SubVolumeGroup CRD
```

weight: 3610
indent: true
---

{% include_relative branch.liquid %}

This guide assumes you have created a Rook cluster as explained in the main [Quickstart guide](quickstart.md)

> # CephFilesystemSubVolumeGroup CRD

Reviewer: Another thought... Do we really need to define a subvolume group in CRDs? Instead, this feels like it could be defined in the storage class. The admin could define the subvolume group name and also quotas in the storage class. This would just require the CSI driver to handle the subvolume group creation the first time it is requested. For RBD namespaces it feels similar where the storage class could define it. However, for external clusters if Rook doesn't have a keyring to create subvolume groups or rbd namespaces, the csi driver approach wouldn't work.

Author: The problem is that the SC is immutable and the quotas are subject to change over time. Also, the CSI driver does not always have the SC details (that's what @Madhu-1 told me). For external mode, the expectation is really to have the subvolumegroup present already. For another external scenario (k8s to k8s), this is handled differently and that's why we would have the CRD.

Reviewer 2: We cannot add the subvolumegroup to the storageclass I explained the problem of adding it to storageclass here https://github.com/ceph/ceph-csi/issues/1884#issuecomment-787665156. This should be added to the csi configmap. Once we have the CRD change we need to work on updating the configmap to have new clusterID as i mentioned here  https://github.com/rook/rook/issues/7036#issuecomment-987568635

Rook allows creation of Ceph Filesystem [SubVolumeGroups](https://docs.ceph.com/en/latest/cephfs/fs-volumes/#fs-subvolume-groups) through the custom resource definitions (CRDs).
Filesystem subvolume groups are an abstraction for a directory level higher than Filesystem subvolumes to effect policies (e.g., File layouts) across a set of subvolumes.
For more information about CephFS volume, subvolumegroup and subvolume refer to the [Ceph docs](https://docs.ceph.com/en/latest/cephfs/fs-volumes/#fs-volumes-and-subvolumes).

## Creating daemon

To get you started, here is a simple example of a CRD to create a subvolumegroup on the CephFilesystem "myfs".

```yaml
apiVersion: ceph.rook.io/v1
kind: CephFilesystemSubVolumeGroup
metadata:

>   name: my-subvolumegroup

Reviewer: What about an "a" suffix for example names instead of the "my" prefix?
```suggestion
  name: group-a
```

  namespace: rook-ceph # namespace:cluster
spec:
  # volumeName is the name of the CephFilesystem CR

>   volumeName: myfs

Reviewer: Since it refers to the Filesystem CR, should we not call it filesystem?
```suggestion
  filesystem: myfs
```

Author: I was hesitant since the official ceph doc refers to it as `volumeName` but filesystem probably makes more sense. Unless we have multiple volume per fs...

Author: I dug a bit more and currently a volume in CephFS is a filesystem, when deploying a CephFilesystem we get one. So no multiple volumes.

```

## Settings

If any setting is unspecified, a suitable default will be used automatically.

### CephFilesystemSubVolumeGroup metadata

- `name`: The name that will be used for the Ceph Filesystem subvolume group.

> - `namespace`: The Kubernetes namespace that will be created for the Rook cluster. The services, pods, and other resources created by the operator will be added to this namespace.

Reviewer: Does this description of the namespace apply? It doesn't seem like any K8s resources will be created in this case

### CephFilesystemSubVolumeGroup spec

> - `volumeName`: The name of the Ceph Filesystem volume.

Reviewer 3: I'm not sure there is a **best** choice for what to name this. If it is `filesystem` or `filesystemName`, it doesn't match Ceph docs as closely. If it's `volumeName`, it doesn't match Rook docs as closely. 

I think it may be slightly easier for users to leave this with the Ceph terminology so that users can glean information more easily from the Ceph docs. However, doing this, I think we should be explicit here in the docs that in Rook, this is the metadata name of a CephFilesystem resource.

Author: I was hesitant to rename this but now that I have a better understanding (hopefully), I feel like we should rename it to `filesystemName` so this matches better with what we have in Rook and like Travis said, explaining what this means in Ceph terminology.

> - `filesystemName`: The name of the Ceph Filesystem volume.

Reviewer: ```suggestion
- `filesystemName`: The metadata name of the CephFilesystem CR where the subvolume group will be created
```
